packages/fee-abstraction/src/storage.rs

```diff
@@ -0,0 +1,388 @@
+/// approval strategy: it always approves `max_fee_amount` for this contract,
+/// overwriting any previous allowance.
+///
+/// # Arguments
+///
+/// * `e` - Access to Soroban environment.
+/// * `fee_token` - The token address to pay the fee with.
+/// * `fee_amount` - The actual fee amount to charge.
+/// * `max_fee_amount` - The maximum fee amount the user approved.
+/// * `expiration_ledger` - The ledger sequence at which the approval expires.
+/// * `user` - The address of the user paying the fee.
+/// * `fee_recipient` - The address that receives the collected fee.
+///
+/// # Events
+///
+/// * topics - `["FeeCollected", user: Address, recipient: Address]`
+/// * data - `[token: Address, amount: i128]`
+///
+/// # Errors
+///
+/// * [`FeeAbstractionError::InvalidFeeBounds`] - If amounts <= 0 or `fee_amount
+///   > max_fee_amount`.
+/// * [`FeeAbstractionError::FeeTokenNotAllowed`] - If the fee token is not
+///   allowed when the allowlist is enabled.
+pub fn collect_fee_with_eager_approval(
+    e: &Env,
+    fee_token: &Address,
+    fee_amount: i128,
+    max_fee_amount: i128,
+    expiration_ledger: u32,
+    user: &Address,
+    fee_recipient: &Address,
+) {
+    check_allowed_fee_token(e, fee_token);
+
+    validate_fee_bounds(e, fee_amount, max_fee_amount);
+
+    let token_client = TokenClient::new(e, fee_token);
+    // User always approves `max_fee_amount` so that the contract can charge up to
+    // that amount, overwriting previous approvals.
+    token_client.approve(user, &e.current_contract_address(), &max_fee_amount, &expiration_ledger);
+
+    token_client.transfer_from(&e.current_contract_address(), user, fee_recipient, &fee_amount);
+
+    emit_fee_collected(e, user, fee_recipient, fee_token, fee_amount);
+}
+
+/// Collect a fee from the user in a given token, overwriting allowance only
+/// when needed.
+///
+/// Compared to [`collect_fee_with_eager_approval`], this variant uses a *lazy*
+/// approval strategy: it approves `max_fee_amount` only if the allowance is
+/// below `max_fee_amount`.
+///
+/// # Arguments
+///
+/// * `e` - Access to Soroban environment.
+/// * `fee_token` - The token address to pay the fee with.
+/// * `fee_amount` - The actual fee amount to charge.
+/// * `max_fee_amount` - The maximum fee amount the user approved.
+/// * `expiration_ledger` - The ledger sequence at which the approval expires.
+/// * `user` - The address of the user paying the fee.
+/// * `fee_recipient` - The address that receives the collected fee.
+///
+/// # Events
+///
+/// * topics - `["FeeCollected", user: Address, recipient: Address]`
+/// * data - `[token: Address, amount: i128]`
+///
+/// # Errors
+///
+/// * [`FeeAbstractionError::InvalidFeeBounds`] - If amounts <= 0 or `fee_amount
+///   > max_fee_amount`.
+/// * [`FeeAbstractionError::FeeTokenNotAllowed`] - If the fee token is not
+///   allowed when the allowlist is enabled.
+pub fn collect_fee_with_lazy_approval(
```

Comment:
too much code duplication across this one and the eager one. 

For the simplest solution that comes to my mind instantly, there can be an internal helper function that can have an `eager` or `lazy` parameter, and branch on the logic accordingly, and these ones will just call that internal function with the parameter. 

And in that case, one can argue, why not make that internal function the public one and get rid of these two. I'm open for both options (keeping these and having an internal helper, or just having one function with eager/lazy with the parameter). 

I'm just pointing out we should make the code more reusable